06_gpu_and_ml/yolo/yolo.py

```diff
@@ -0,0 +1,288 @@
+
+def convert_dataset_to_yolo(dataset_path, output_path):
+    from tqdm import tqdm
+    root_folder = Path(output_path)
+    if not root_folder.exists():
+        root_folder.mkdir(parents=True)
+
+    category_map = {
+                    "0": "Caption", "1": "Footnote", "2": "Formula", "3": "List-item",
+                    "4": "Page-footer", "5": "Page-header", "6": "Picture",
+                    "7": "Section-header", "8": "Table", "9": "Text", "10": "Title",
+                }
+    inverse_category_map = {v: k for k, v in category_map.items()}
+    with open(root_folder / "data.yaml", "w") as f:
+        yaml.dump(
+            {
+                "path": root_folder.absolute().as_posix(),
+                "train": "./images/train",
+                "val": "./images/val",
+                "test": "./images/test",
+                "names": category_map,
+            },
+            f,
+        )
+ 
+    for split in ["train", "val", "test"]:
+        print(f"Converting {split} dataset...")
+        
+        ann_dir = os.path.join(dataset_path, f"base_dataset/{split}/annotations")
+        img_dir = os.path.join(dataset_path, f"base_dataset/{split}/images")
+        
+        os.makedirs(root_folder / "labels" / split, exist_ok=True)
+        os.makedirs(root_folder / "images" / split, exist_ok=True)
+
+        for file in tqdm(os.listdir(ann_dir)):
+            file_path = os.path.join(ann_dir, file)
+            with open(file_path, "r", encoding="utf8") as f:
+                data = json.load(f)
+
+            image_id = os.path.splitext(file)[0]
+            src_image_path = os.path.join(img_dir, f"{image_id}.png")
+            dst_image_path = root_folder / "images" / split / f"{image_id}.png"
+            # Have to copy this is cross device
+            shutil.copy(src_image_path, dst_image_path)
+
+            coco_width, coco_height = data["metadata"]["coco_width"], data["metadata"]["coco_height"]
+            
+            with open(root_folder / "labels" / split / f"{image_id}.txt", "w") as f:
+                seen_annotations = set()
+                for item in data["form"]:
+                    category_str = item["category"]
+                    category = inverse_category_map[category_str]
+
+                    bbox = item["box"]
+                    left, top, width, height = bbox
+                    left /= coco_width
+                    top /= coco_height
+                    width /= coco_width
+                    height /= coco_height
+                    center_x = left + width / 2
+                    center_y = top + height / 2
+
+                    annotation = (category, center_x, center_y, width, height)
+                    if annotation not in seen_annotations:
+                        f.write(f"{category} {center_x} {center_y} {width} {height}\n")
+                        seen_annotations.add(annotation)
+
+@app.function(
+    image=yolo_image,
+    volumes={"/root/data/": volume},
+    cpu=4,
```

Comment:
We generally don't request lots of CPU on our examples, because jobs can usually burst up to consume more CPU when they need it. Did you find that this example required CPU pinning for some reason?